.github/workflows/ci.yml

```diff
@@ -0,0 +1,97 @@
+name: CI
+
+on:
+  pull_request:
+    types: [opened, synchronize]
+
+jobs:
+  test-on-ubuntu:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-18.04, ubuntu-20.04]
+        compiler: [{C: gcc-7, CXX: g++-7}, {C: gcc-11, CXX: g++-11}]
+        python: [python3.8, python3.9, python3.10]
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      # - Installing software-properties-common adds add-apt-repository
+      - name: Update install tools
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y software-properties-common curl
+
+      # - ppa:ubuntu-toolchain-r/test is needed to have compiler available on all ubuntu versions
+      # - Installing g++-x also installs gcc-x
+      - name: Install C and CXX compilers
+        run: |
+          sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
+          sudo apt-get install -y ${{ matrix.compiler.CXX }}
+
+      # - ppa:deadsnakes/ppa is needed to have python3.x-dev available on all ubuntu versions
+      # - python3.x-dev is used instead of python3.x because the -dev version includes python header
+      # files needed to compile C extension modules
+      - name: Install python
+        run: |
+          sudo add-apt-repository -y ppa:deadsnakes/ppa
+          sudo apt-get install -y ${{ matrix.python }}-dev ${{ matrix.python }}-distutils
+          curl https://bootstrap.pypa.io/get-pip.py | ${{ matrix.python }}
```

Comment:
Is there a reason to install pip this way instead of just installing `${{ matrix.python }}-pip`?

Reply:
Okay. As we discussed in person, one can't `apt-get install python3.10-pip`, as it doesn't seem to exist. One can install `apt-get install python3-pip`, but that just installs the pip for the system python (which doesn't play nicely with our non-system python).

Downloading from `https://bootstrap.pypa.io/get-pip.py` gets us the latest pip and always works. If we are worried about breaking on older versions of pips, the pip installed by that script can downgrade itself, so we can just set it to whatever version we'd like.